openssl/ecdh.go

```diff
@@ -14,50 +14,65 @@ import (
 )
 
 type PublicKeyECDH struct {
-	curve string
-	key   C.GO_EC_POINT_PTR
+	_pkey C.GO_EVP_PKEY_PTR
 	bytes []byte
 }
 
 func (k *PublicKeyECDH) finalize() {
-	C.go_openssl_EC_POINT_free(k.key)
+	C.go_openssl_EVP_PKEY_free(k._pkey)
 }
 
 type PrivateKeyECDH struct {
-	curve string
-	key   C.GO_EC_KEY_PTR
+	_pkey C.GO_EVP_PKEY_PTR
 }
 
 func (k *PrivateKeyECDH) finalize() {
-	C.go_openssl_EC_KEY_free(k.key)
+	C.go_openssl_EVP_PKEY_free(k._pkey)
 }
 
 func NewPublicKeyECDH(curve string, bytes []byte) (*PublicKeyECDH, error) {
 	if len(bytes) < 1 {
 		return nil, errors.New("NewPublicKeyECDH: missing key")
 	}
-
 	nid, err := curveNID(curve)
 	if err != nil {
 		return nil, err
 	}
-
-	group := C.go_openssl_EC_GROUP_new_by_curve_name(nid)
-	if group == nil {
-		return nil, newOpenSSLError("EC_GROUP_new_by_curve_name")
-	}
-	defer C.go_openssl_EC_GROUP_free(group)
-	key := C.go_openssl_EC_POINT_new(group)
+	key := C.go_openssl_EC_KEY_new_by_curve_name(nid)
 	if key == nil {
-		return nil, newOpenSSLError("EC_POINT_new")
+		return nil, newOpenSSLError("EC_KEY_new_by_curve_name")
 	}
-	ok := C.go_openssl_EC_POINT_oct2point(group, key, base(bytes), C.size_t(len(bytes)), nil) != 0
-	if !ok {
-		C.go_openssl_EC_POINT_free(key)
-		return nil, errors.New("point not on curve")
+	var k *PublicKeyECDH
+	defer func() {
+		if k == nil {
+			C.go_openssl_EC_KEY_free(key)
+		}
+	}()
+	if vMajor == 1 && vMinor == 0 {
+		// EC_KEY_oct2key does not exist on OpenSSL 1.0.2,
+		// we have to simulate it.
+		group := C.go_openssl_EC_KEY_get0_group(key)
+		pt := C.go_openssl_EC_POINT_new(group)
+		if pt == nil {
+			return nil, newOpenSSLError("EC_POINT_new")
+		}
+		defer C.go_openssl_EC_POINT_free(pt)
```

Comment:
Does freeing `pt` here also free `group`?

Reply:
Nop, and there is no need to free `group`, as it will be freed when the key that owns is is freed.